pinot-server/src/main/java/org/apache/pinot/server/starter/helix/SendStatsPredicate.java

```diff
@@ -0,0 +1,160 @@
+        return new Safe();
+      }
+    },
+    ALWAYS {
+      @Override
+      public SendStatsPredicate create() {
+        return new SendStatsPredicate() {
+          @Override
+          public boolean getSendStats() {
+            return false;
+          }
+
+          @Override
+          public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+            // Nothing to do
+          }
+        };
+      }
+    },
+    NEVER {
+      @Override
+      public SendStatsPredicate create() {
+        return new SendStatsPredicate() {
+          @Override
+          public boolean getSendStats() {
+            return false;
+          }
+
+          @Override
+          public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+            // Nothing to do
+          }
+        };
+      }
+    };
+
+    public abstract SendStatsPredicate create();
+  }
+
+  private static class Safe extends SendStatsPredicate {
+    private final AtomicBoolean _sendStats = new AtomicBoolean(true);
+
+    @Override
+    public boolean getSendStats() {
+      return _sendStats.get();
+    }
+
+    @Override
+    public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+      boolean sendStats = true;
+      // TODO: Right now we read all configs, while only broker and servers are actually important
+      //  But it isn't clear how to detect for sure that an instance config belongs to a broker or server
+      for (InstanceConfig instanceConfig : instanceConfigs) {
+        String otherVersion = instanceConfig.getRecord()
+            .getStringField(CommonConstants.Helix.Instance.PINOT_VERSION_KEY, null);
+        if (otherVersion == null || otherVersion.equals(PinotVersion.UNKNOWN)) {
+          LOGGER.warn("Instance {} does not have version", instanceConfig.getInstanceName());
+          continue;
+        }
+        if (isProblematicVersion(otherVersion)) {
+          LOGGER.info("Found problematic version {} on instance {}. Stats may not be recognized", otherVersion,
+              instanceConfig.getInstanceName());
+          sendStats = false;
+          break;
+        }
+      }
+      if (_sendStats.getAndSet(sendStats) != sendStats) {
+        LOGGER.warn("Send MSE stats is now {}", sendStats ? "enabled" : "disabled");
+      }
+    }
+
+    /// Returns true if the version is problematic
+    ///
+    /// Ideally [PinotVersion] should have a way to extract versions in comparable format, but given it doesn't we
+    /// need to parse the string here. In case version doesn't match `1\.x\..*`, we treat is as a problematic version
+    private boolean isProblematicVersion(@Nullable String versionStr) {
+      if (versionStr == null) {
+        return true;
+      }
+      if (versionStr.equals(PinotVersion.UNKNOWN)) {
+        return true;
+      }
+      // Lets try to parse 1.x versions
```

Comment:
Do we need the following check? If the version is present, it is safe version right?

Reply:
I'm just being extra careful, given this code is sensitive. We don't want to throw an array out of bound exception here just because for whatever reason we change the version format to something that, for example, doesn't include dots

Reply:
I'm also adding a check to compare the value with the current version of the process. Given most of the cluster will be homogeneous, most of the time, it won't be necessary to parse the version